### ISSUE TITLE
Driver / Platform Speed Evaluations

### ISSUE THREAD

**Issue by wanji** [BVLC/caffe#579]
(This issue is related to #388, #483 and #574.)

I have found that the backward of convolutional layer is very slow and made some tests on image-net dataset. Hope this can help figuring out the problem which slow down the backward procedure.
##### The results on CentOS and Ubuntu are quite different.

Add the following timing code to `ConvolutionLayer<Dtype>::Backward_gpu`, in which `toc()` return the time elapsed from `tic()` in milliseconds:

```
tic();
double time[] = {0, 0, 0, 0, 0};
  for (int n = 0; n < num_; ++n) {
time[0] += toc(); tic();
    // since we saved memory in the forward pass by not storing all col data,
    // we will need to recompute them.
    im2col_gpu(bottom_data + (*bottom)[0]->offset(n), channels_, height_,
                      width_, kernel_size_, pad_, stride_, col_data);
time[1] += toc(); tic();
    // gradient w.r.t. weight. Note that we will accumulate diffs.
    for (int g = 0; g < group_; ++g) {
      caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasTrans, M_, K_, N_,
        (Dtype)1., top_diff + top[0]->offset(n) + top_offset * g,
        col_data + col_offset * g, (Dtype)1.,
        weight_diff + weight_offset * g);
    }
time[2] += toc(); tic();
    // gradient w.r.t. bottom data, if necessary
    if (propagate_down) {
      for (int g = 0; g < group_; ++g) {
        caffe_gpu_gemm<Dtype>(CblasTrans, CblasNoTrans, K_, N_, M_,
          (Dtype)1., weight + weight_offset * g,
          top_diff + top[0]->offset(n) + top_offset * g,
          (Dtype)0., col_diff + col_offset * g);
      }
time[3] += toc(); tic();
      // col2im back to the data
      col2im_gpu(col_diff, channels_, height_, width_, kernel_size_, pad_,
          stride_, bottom_diff + (*bottom)[0]->offset(n));
time[4] += toc(); tic();
    }
  }
LOG(INFO) << "\t" << this->layer_param_.name() << "(gpu) - end"
  << "\t" << time[0]
  << "\t" << time[1]
  << "\t" << time[2]
  << "\t" << time[3]
  << "\t" << time[4];
```
#### CentOS 6.4, K20, Driver Version 331.75. (CUDA 5.0/5.5/6.0 got similar results):

```
I0702 11:36:11.680141  9021 conv_layer.cu:122]  conv5(gpu) - end        0.022   3.52    6.606   377.492 62.175
I0702 11:36:12.915659  9021 conv_layer.cu:122]  conv4(gpu) - end        0.017   1.951   3.49    1054.78 173.866
I0702 11:36:14.315760  9021 conv_layer.cu:122]  conv3(gpu) - end        0.03    1.967   562.029 559.671 274.977
I0702 11:36:16.979482  9021 conv_layer.cu:122]  conv2(gpu) - end        0.021   1.951   3.621   2280.79 375.892
I0702 11:36:18.843824  9021 conv_layer.cu:122]  conv1(gpu) - end        0.02    2.021   1.988   1398.12 460.728
```
#### Ubuntu 12.04, GTX680, Driver Version 331.62. (CUDA 6.0):

```
I0702 11:35:54.877564  4720 conv_layer.cu:122]  conv5(gpu) - end        0.008   14.443  58.896  50.811  5.854
I0702 11:35:55.035778  4720 conv_layer.cu:122]  conv4(gpu) - end        0.007   0.947   79.334  63.432  5.967
I0702 11:35:55.168871  4720 conv_layer.cu:122]  conv3(gpu) - end        0.009   45.001  47.765  22.549  5.445
I0702 11:35:55.490033  4720 conv_layer.cu:122]  conv2(gpu) - end        0.007   4.711   149.377 126.984 19.074
I0702 11:35:55.957451  4720 conv_layer.cu:122]  conv1(gpu) - end        0.005   1.183   432.66  11.169  1.415
```
##### The time consumed by GPU are similar in the two configurations (around 2 second).

Add `exit(0);` after the first `Dtype loss = net_->ForwardBackward(bottom_vec);` in `Solver<Dtype>::Solve`, and using nvprof for evaluation:
#### CentOS 6.4, K20, Driver Version 331.75. (CUDA 6.0):

```
Time(%)      Time     Calls       Avg       Min       Max  Name
 13.35%  265.08ms      1024  258.87us  202.06us  321.74us  void sgemm_largek_lds64<bool=0, bool=0, int=5, int=5, int=4, int=4, int=4, int=32>(float*, float const *, float const *, int, int, int, int, int, int, float const *, float const *, float, float, int, int, int*, int*)
 12.13%  240.91ms      1024  235.26us  182.06us  366.54us  sgemm_sm35_ldg_nn_64x16x64x16x16
 12.00%  238.36ms      1536  155.18us  100.26us  226.25us  sgemm_sm35_ldg_tn_32x16x64x8x16
 10.53%  209.09ms      2048  102.09us  57.890us  171.08us  sgemm_sm35_ldg_nt_128x8x128x16x16
  7.11%  141.19ms       257  549.37us  509.91us  4.0105ms  void sgemm_largek_lds64<bool=1, bool=0, int=5, int=5, int=4, int=4, int=4, int=34>(float*, float const *, float const *, int, int, int, int, int, int, float const *, float const *, float, float, int, int, int*, int*)
  6.58%  130.65ms      2560  51.035us  21.537us  103.81us  void caffe::im2col_gpu_kernel<float>(int, float const *, int, int, int, int, int, int, int, caffe::im2col_gpu_kernel<float>*)
  6.11%  121.32ms      2048  59.235us  43.266us  109.80us  sgemm_sm35_ldg_nt_64x16x128x8x32
  5.67%  112.59ms        29  3.8823ms  1.0880us  44.767ms  [CUDA memcpy HtoD]
  5.07%  100.64ms      1795  56.065us  44.674us  269.58us  sgemm_sm35_ldg_nt_128x16x64x16x16
  4.32%  85.750ms      1280  66.991us  34.273us  106.05us  void caffe::col2im_gpu_kernel<float>(int, float const *, int, int, int, int, int, int, int, int, caffe::col2im_gpu_kernel<float>*)
  4.02%  79.919ms       258  309.76us  249.10us  10.962ms  sgemm_sm35_ldg_tn_128x8x256x16x32
  2.57%  50.944ms         3  16.981ms  4.2624ms  28.305ms  void caffe::MaxPoolBackward<float>(int, float const *, float const , float const , int, int, int, int, int, int, int, int, caffe::MaxPoolBackward<float>*)
  1.71%  33.874ms      1281  26.443us  8.5120us  81.892us  void gemv2T_kernel_val<float, int=128, int=16, int=2, int=2, bool=0>(int, int, float, float const *, int, float const *, int, float, float*, int)
  1.70%  33.665ms       256  131.50us  123.11us  149.57us  sgemm_sm35_ldg_tn_64x16x128x8x32
  1.11%  21.983ms      1351  16.271us  3.9360us  1.9074ms  [CUDA memset]
  0.76%  15.124ms      1284  11.778us  6.5280us  61.763us  void gemmk1_kernel<float, int=256, int=5, bool=0, bool=0, bool=0, bool=0>(cublasGemmk1Params<float>, float const *, float const *, float*)
  0.76%  15.039ms         3  5.0131ms  1.1581ms  9.5665ms  sgemm_sm_heavy_nt_ldg
  0.72%  14.347ms         7  2.0496ms  91.875us  6.3059ms  void caffe::ReLUBackward<float>(int, float const *, float const , caffe::ReLUBackward<float>*)
  0.72%  14.334ms         2  7.1668ms  6.0540ms  8.2797ms  void caffe::LRNComputeDiff<float>(int, float const *, float const , float const , float const , int, int, int, int, int, caffe::LRNComputeDiff<float>, caffe::LRNComputeDiff<float>, caffe::LRNComputeDiff<float>*)
  0.57%  11.394ms         3  3.7979ms  1.0240ms  6.0163ms  void caffe::MaxPoolForward<float>(int, float const *, int, int, int, int, int, int, int, int, caffe::MaxPoolForward<float>*)
  0.56%  11.155ms         3  3.7184ms  1.0767ms  6.1215ms  sgemm_sm_heavy_nn_ldg
  0.53%  10.604ms         1  10.604ms  10.604ms  10.604ms  generate_seed_pseudo(__int64, __int64, __int64, curandOrdering, curandStateXORWOW*, unsigned int*)
  0.52%  10.231ms         7  1.4616ms  65.378us  4.5021ms  void caffe::ReLUForward<float>(int, float const *, caffe::ReLUForward<float>*)
  0.23%  4.5314ms         3  1.5105ms  433.14us  2.4654ms  sgemm_sm35_ldg_nn_128x16x64x16x16
  0.18%  3.6058ms         2  1.8029ms  1.7929ms  1.8129ms  void caffe::LRNFillScale<float>(int, float const *, int, int, int, int, int, caffe::LRNFillScale<float>, caffe::LRNFillScale<float>*)
  0.17%  3.4028ms         2  1.7014ms  1.3006ms  2.1022ms  void caffe::LRNComputeOutput<float>(int, float const *, float const , caffe::LRNComputeOutput<float>, caffe::LRNComputeOutput<float>*)
  0.14%  2.8790ms      1025  2.8080us  2.2720us  6.9770us  void scal_kernel<float, int=1, bool=1, int=6, int=5, int=5, int=3>(cublasTransposeParams<float>, float const *, float*, float const *)
  0.08%  1.6080ms       256  6.2810us  6.0160us  6.4960us  void scal_kernel<float, int=1, bool=0, int=6, int=5, int=5, int=3>(cublasTransposeParams<float>, float const *, float*, float const *)
  0.04%  822.50us         1  822.50us  822.50us  822.50us  void caffe::kernel_get_max<float>(int, int, float const *, caffe::kernel_get_max<float>*)
  0.01%  201.10us         2  100.55us  100.26us  100.84us  void caffe::DropoutForward<float>(int, float const *, unsigned int const *, unsigned int, float, caffe::DropoutForward<float>*)
  0.01%  198.18us         2  99.092us  99.076us  99.108us  void caffe::DropoutBackward<float>(int, float const *, unsigned int const *, unsigned int, float, caffe::DropoutBackward<float>*)
  0.01%  158.66us         2  79.331us  3.2320us  155.43us  [CUDA memcpy DtoH]
  0.01%  119.17us         2  59.586us  58.915us  60.258us  void gen_sequenced<curandStateXORWOW, unsigned int, int, __operator_&__(unsigned int curand_noargs<curandStateXORWOW>(curandStateXORWOW*, int))>(curandStateXORWOW*, unsigned int*, unsigned long, unsigned long, int)
  0.01%  114.21us         2  57.106us  56.034us  58.179us  void gemv2N_kernel_val<float, int=128, int=4, int=4, int=4>(float, float, cublasGemv2Params<float>)
  0.00%  22.145us         1  22.145us  22.145us  22.145us  void caffe::kernel_softmax_div<float>(int, int, float const *, caffe::kernel_softmax_div<float>*)
  0.00%  19.457us         1  19.457us  19.457us  19.457us  [CUDA memcpy DtoD]
  0.00%  18.944us         1  18.944us  18.944us  18.944us  void caffe::kernel_exp<float>(int, float const *, caffe::kernel_exp<float>*)
  0.00%  18.497us         1  18.497us  18.497us  18.497us  void gemv2N_kernel_val<float, int=128, int=8, int=4, int=4>(float, float, cublasGemv2Params<float>)
```
#### Ubuntu 12.04, GTX680, Driver Version 331.62. (CUDA 6.0):

```
Time(%)      Time     Calls       Avg       Min       Max  Name
 19.99%  367.53ms      1795  204.75us  92.642us  14.241ms  sgemm_sm30_ldg_tex_tn_64x16x128x16x32
 19.17%  352.42ms      2050  171.91us  85.410us  13.351ms  sgemm_sm30_ldg_tex_nt_64x16x128x16x32
 17.76%  326.54ms      1539  212.18us  161.54us  11.779ms  sgemm_sm30_ldg_tex_nn_64x16x64x16x16
 11.15%  205.07ms       255  804.18us  760.49us  848.43us  void sgemm_largek_lds64<bool=1, bool=0, int=5, int=5, int=4, int=4, int=4, int=34>(float*, float const *, float const *, int, int, int, int, int, int, float const *, float const *, float, float, int, int, int*, int*)
  6.80%  124.94ms       512  244.03us  222.60us  274.85us  void sgemm_largek_lds64<bool=0, bool=0, int=5, int=5, int=4, int=4, int=4, int=32>(float*, float const *, float const *, int, int, int, int, int, int, float const *, float const *, float, float, int, int, int*, int*)
  6.79%  124.84ms      2559  48.785us  20.609us  103.59us  void caffe::im2col_gpu_kernel<float>(int, float const *, int, int, int, int, int, int, int, caffe::im2col_gpu_kernel<float>*)
  4.67%  85.855ms      1279  67.126us  32.129us  107.33us  void caffe::col2im_gpu_kernel<float>(int, float const *, int, int, int, int, int, int, int, int, caffe::col2im_gpu_kernel<float>*)
  3.68%  67.623ms        29  2.3318ms     768ns  26.544ms  [CUDA memcpy HtoD]
  2.90%  53.256ms         3  17.752ms  4.4387ms  29.608ms  void caffe::MaxPoolBackward<float>(int, float const *, float const , float const , int, int, int, int, int, int, int, int, caffe::MaxPoolBackward<float>*)
  1.20%  22.105ms      1281  17.256us  5.4720us  53.345us  void gemv2T_kernel_val<float, int=128, int=16, int=2, int=2, bool=0>(int, int, float, float const *, int, float const *, int, float, float*, int)
  1.06%  19.473ms       838  23.237us  2.2400us  2.0989ms  [CUDA memset]
  0.85%  15.629ms         2  7.8143ms  6.6054ms  9.0232ms  void caffe::LRNComputeDiff<float>(int, float const *, float const , float const , float const , int, int, int, int, int, caffe::LRNComputeDiff<float>, caffe::LRNComputeDiff<float>, caffe::LRNComputeDiff<float>*)
  0.78%  14.307ms      1284  11.142us  4.6400us  58.657us  void gemmk1_kernel<float, int=256, int=5, bool=0, bool=0, bool=0, bool=0>(cublasGemmk1Params<float>, float const *, float const *, float*)
  0.71%  13.056ms         7  1.8651ms  82.305us  5.7611ms  void caffe::ReLUBackward<float>(int, float const *, float const , caffe::ReLUBackward<float>*)
  0.70%  12.950ms         3  4.3165ms  1.0945ms  7.0842ms  void caffe::MaxPoolForward<float>(int, float const *, int, int, int, int, int, int, int, int, caffe::MaxPoolForward<float>*)
  0.59%  10.784ms         1  10.784ms  10.784ms  10.784ms  generate_seed_pseudo(__int64, __int64, __int64, curandOrdering, curandStateXORWOW*, unsigned int*)
  0.56%  10.316ms         7  1.4738ms  63.842us  4.6414ms  void caffe::ReLUForward<float>(int, float const *, caffe::ReLUForward<float>*)
  0.25%  4.5825ms         2  2.2912ms  2.0010ms  2.5814ms  void caffe::LRNFillScale<float>(int, float const *, int, int, int, int, int, caffe::LRNFillScale<float>, caffe::LRNFillScale<float>*)
  0.22%  4.0082ms         2  2.0041ms  1.4767ms  2.5315ms  void caffe::LRNComputeOutput<float>(int, float const *, float const , caffe::LRNComputeOutput<float>, caffe::LRNComputeOutput<float>*)
  0.06%  1.0850ms       255  4.2540us  4.0960us  4.4160us  void scal_kernel<float, int=1, bool=0, int=6, int=5, int=5, int=3>(cublasTransposeParams<float>, float const *, float*, float const *)
  0.05%  873.93us       512  1.7060us  1.3760us  2.0160us  void scal_kernel<float, int=1, bool=1, int=6, int=5, int=5, int=3>(cublasTransposeParams<float>, float const *, float*, float const *)
  0.03%  598.89us         1  598.89us  598.89us  598.89us  void caffe::kernel_get_max<float>(int, int, float const *, caffe::kernel_get_max<float>*)
  0.01%  185.09us         2  92.546us  92.482us  92.610us  void caffe::DropoutForward<float>(int, float const *, unsigned int const *, unsigned int, float, caffe::DropoutForward<float>*)
  0.01%  184.45us         2  92.226us  92.194us  92.258us  void caffe::DropoutBackward<float>(int, float const *, unsigned int const *, unsigned int, float, caffe::DropoutBackward<float>*)
  0.01%  157.25us         2  78.625us  2.3040us  154.95us  [CUDA memcpy DtoH]
  0.00%  79.201us         2  39.600us  39.552us  39.649us  void gen_sequenced<curandStateXORWOW, unsigned int, int, __operator_&__(unsigned int curand_noargs<curandStateXORWOW>(curandStateXORWOW*, int))>(curandStateXORWOW*, unsigned int*, unsigned long, unsigned long, int)
  0.00%  72.609us         2  36.304us  36.160us  36.449us  void gemv2N_kernel_val<float, int=128, int=4, int=4, int=4>(float, float, cublasGemv2Params<float>)
  0.00%  24.288us         1  24.288us  24.288us  24.288us  void caffe::kernel_softmax_div<float>(int, int, float const *, caffe::kernel_softmax_div<float>*)
  0.00%  22.176us         1  22.176us  22.176us  22.176us  void caffe::kernel_exp<float>(int, float const *, caffe::kernel_exp<float>*)
  0.00%  16.705us         1  16.705us  16.705us  16.705us  [CUDA memcpy DtoD]
  0.00%  13.120us         1  13.120us  13.120us  13.120us  void gemv2N_kernel_val<float, int=128, int=8, int=4, int=4>(float, float, cublasGemv2Params<float>)
```
